frontend-guide.md

```diff
@@ -0,0 +1,333 @@
+The build usually includes:
+- Converting SASS to CSS
+- Bundling CSS
+- Versionifying CSS bundle
+- Running Flow type check
+- Running ESLint
+- Bundling JS
+- Uglyfying JS bundle
+- Versionifying JS bundle
+- Bundling JS vendor files
+- Uglyfying JS vendor bundle
+- Versionifying JS vendor bundle
+- Running the test
+
+### Setting the project
+
+We have multiple scripts to help setting the frontend.
+
+**Main setup all**
+
+```
+cd megarepo/frontend
+./setup.sh
+```
+
+This will install all the dependencies for all project without building each one of them.
+
+**Main build all**
+
+```
+cd megarepo/frontend
+./build-all.sh
+```
+
+This will run the build for each individual project but will assume the project was setup successfully. It may take some time.
+
+**Reset all**
+
+```
+cd megarepo/frontend
+./reset-all.sh
+```
+
+This will remove the `node_modules` folder of each project, re-install AND build every project individually. This is useful when something went wrong in your repos and you need a clean blank state.
+
+### Running the build
+
+To run the build or part of the build, we use NPM scripts. They are the same for all web applications.
+
+**Build for dev environment**
+```
+yarn build:dev
+```
+
+**Build for prod environment**
+```
+yarn build:prod
+or
+yarn build
+```
+
+**Launch the watcher and rebuild for dev after each local change**
+```
+yarn watch
+```
+
+**Run ESLint with all warnings**
+```
+yarn lint
+```
+
+### Internal dependencies
+
+In order to share code of `frontend-entities` between multiple projects, we import the folder as a package.
+The build process
+
+
+### Main tools used for the build
+
+
+#### Browserify
+
+We use Browserify to create JS bundles of our code and vendor files. We use a lib on top of Browserify called [`browserify-incremental`](https://github.com/jsdf/browserify-incremental) allowing us to speed us the bundling process by not re-creating the bundle from scratch after every change.
+Each build type (`dev` or `prod`) use a different cache file for this process (`browserify-cache-development.json`).
+Browserify is supplemented by [`babelify`](https://github.com/babel/babelify) which compile our ES6 code into browser compatible Javascript (learn more about [Babel](http://babeljs.io/]).
+
+#### Flow
+
+See Stack/Flow
+
+
+## Stack
+
+The stack of the frontend is almost the same for each project with few exceptions.
+
+### Backbone Collections and Models
+
+When calling an API resource, we use Backbone Collections and Models to transform them into usable, type-checked, javascript usuable class.
+Every resource is atomically defined by it's model. Each model contains a `parse` function validating and transforming data from API response.
+
+_Example of the model parse function_
+```javascript
+parse: function(response: Object): Object {
+  // Validating the API response contains the keys we expect and the values are the type we expect
```

Comment:
"Validating the API response contains the keys and values we expect _at runtime_"